test/src/ABE.test.c

```diff
@@ -0,0 +1,136 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "gmp.h"
+
+#include "greatest.h"
+
+#include "attribute-based/ABE.h"
+
+TEST basic_abe_test(SecurityLevel securityLevel, char* message, AccessTree* accessTree, char** attributes, int expectedReponse)
+{
+    PublicKey_ABE* publickey = malloc(sizeof (PublicKey_ABE));
+    MasterKey_ABE* masterkey = malloc(sizeof (MasterKey_ABE));
+    CryptidStatus status = cryptid_setup_ABE(securityLevel, publickey, masterkey);
+    ASSERT_EQ(status, CRYPTID_SUCCESS);
+
+
+    EncryptedMessage_ABE* encrypted = malloc(sizeof (EncryptedMessage_ABE));
+    status = cryptid_encrypt_ABE(encrypted, message, strlen(message), publickey, accessTree);
+    ASSERT_EQ(status, CRYPTID_SUCCESS);
+
+    SecretKey_ABE* secretkey = malloc(sizeof (SecretKey_ABE));
+
+    status = cryptid_keygen_ABE(masterkey, attributes, secretkey);
+    ASSERT_EQ(status, CRYPTID_SUCCESS);
+
+    SecretKey_ABE* secretkey_new = malloc(sizeof (SecretKey_ABE));
+
+    status = cryptid_delegate_ABE(secretkey, attributes, secretkey_new);
+    ASSERT_EQ(status, CRYPTID_SUCCESS);
+
+    char* result_new;
+    CryptidStatus status_new = cryptid_decrypt_ABE(&result_new, encrypted, secretkey_new);
+    ASSERT_EQ(status, CRYPTID_SUCCESS);
+    char* result;
+    status = cryptid_decrypt_ABE(&result, encrypted, secretkey);
+
+    if(expectedReponse == 1)
+    {
+        ASSERT_EQ(status_new, CRYPTID_SUCCESS);
+        ASSERT_EQ(status, CRYPTID_SUCCESS);
+        ASSERT_EQ(strcmp(result, message), 0);
+        ASSERT_EQ(strcmp(result_new, message), 0);
+
+        free(result);
+        free(result_new);
+    }
+    else
+    {
+        ASSERT_EQ(status_new, CRYPTID_ILLEGAL_PRIVATE_KEY_ERROR);
+        ASSERT_EQ(status, CRYPTID_ILLEGAL_PRIVATE_KEY_ERROR);
+    }
+
+    destroyPublicKey_ABE(publickey);
+    destroyMasterKey_ABE(masterkey);
+    destroyEncryptedMessage_ABE(encrypted);
+    destroySecretKey_ABE(secretkey);
+    destroySecretKey_ABE(secretkey_new);
+
+    PASS();
+}
+
+static void generateRandomString(char** output, size_t outputLength, char* alphabet, size_t alphabetSize)
+{
+    memset(*output, '\0', outputLength);
+    
+    for (size_t i = 0; i < outputLength - 1; i++)
+    {
+        size_t index = rand() % (alphabetSize - 1);
+
+        (*output)[i] = alphabet[index];
+    }
+
+    (*output)[outputLength - 1] = '\0';
+}
+
+SUITE(cryptid_abe_suite)
```

Comment:
I'm a bit concerned regarding the testing strategy here. What do you think, what would be the best strategy to cover most of the cases possible?

In my opinion, instead of a randomized test, we should opt here for targeted tests, based on the access tree. In detail, I think, that we should test with many different access tree including different node combinations (OR, AND), just to see, whether it handles all cases correctly.